signer/src/fees.rs

```diff
@@ -0,0 +1,355 @@
+    ) -> impl Future<Output = Result<FeeEstimate, Error>> + Send;
+}
+
+impl FeeEstimator for BitcoinerLive {
+    /// Fetch the fee estimate from bitcoiner.live.
+    ///
+    /// The returned value gives a fee estimate where there is a 90%
+    /// probability that the transaction will be confirmed within 30
+    /// minutes.
+    async fn estimate_fee_rate(&self, client: &reqwest::Client) -> Result<FeeEstimate, Error> {
+        let url = format!(
+            "{}/api/fees/estimates/latest?confidence=0.9",
+            &self.base_url
+        );
+        let resp: BitcoinerLiveResponse = client
+            .get(url)
+            .timeout(REQUEST_TIMEOUT)
+            .send()
+            .await?
+            .json()
+            .await?;
+
+        let now = time::OffsetDateTime::now_utc().unix_timestamp();
+        if now - resp.timestamp > FIVE_MINUTES_SECONDS {
+            return Err(Error::OldFeeEstimate);
+        }
+
+        Ok(FeeEstimate {
+            sats_per_vbyte: resp.estimates.thirty.sat_per_vbyte,
+        })
+    }
+}
+
+impl FeeEstimator for MempoolSpace {
+    /// Fetch the fee estimate from mempool.space
+    ///
+    /// The returned value is the High Priority fee rate displayed on
+    /// https://mempool.space.
+    async fn estimate_fee_rate(&self, client: &reqwest::Client) -> Result<FeeEstimate, Error> {
+        let url = format!("{}/api/v1/fees/recommended", &self.base_url);
+        let resp: MempoolSpaceResponse = client
+            .get(url)
+            .timeout(REQUEST_TIMEOUT)
+            .send()
+            .await?
+            .json()
+            .await?;
+
+        Ok(FeeEstimate {
+            sats_per_vbyte: resp.fastest_fee as f64,
+        })
+    }
+}
+
+#[derive(Debug)]
+enum FeeSource {
+    BitcoinerLive(BitcoinerLive),
+    MempoolSpace(MempoolSpace),
+}
+
+impl FeeEstimator for FeeSource {
+    async fn estimate_fee_rate(&self, client: &reqwest::Client) -> Result<FeeEstimate, Error> {
+        match self {
+            Self::BitcoinerLive(btclive) => btclive.estimate_fee_rate(client).await,
+            Self::MempoolSpace(mempool) => mempool.estimate_fee_rate(client).await,
+        }
+    }
+}
+
+/// Compute the average price of the fee estimates from the given sources.
+async fn estimate_fee_rate_impl<T>(sources: &[T], client: &reqwest::Client) -> Result<f64, Error>
+where
+    T: FeeEstimator,
+{
+    let futures_iter = sources
+        .iter()
+        .map(|source| async move { source.estimate_fee_rate(client).await });
+    let mut responses = futures::future::join_all(futures_iter).await;
+
+    if responses.iter().all(Result::is_err) {
+        return Err(Error::NoGoodFeeEstimates);
+    }
+
+    responses.retain(Result::is_ok);
+    let num_responses = responses.len();
+    let sum_sats_per_vbyte = responses
+        .into_iter()
+        .filter_map(Result::ok)
+        .map(|x| x.sats_per_vbyte)
+        .sum::<f64>();
+
+    Ok(sum_sats_per_vbyte / num_responses as f64)
+}
```

Comment:
This feels like the functionality of some type of aggregate fee rate estimator. May I suggest implementing a `AverageFeeRate` type in the style of:

```
struct AverageFeeRate<T> {
    sources: Vec<T>, // This could be a [T; N] if we want to use const generics and skip the heap allocation
}

impl<T: FeeEstimator> FeeEstimator for AverageFeeRate<T> {
    async fn estimate_fee_rate(...) -> ...
}
```

Reply:
Yeah I like this approach

Reply:
Okay, I briefly tried out the `AverageFeeRate` approach but ended up removing it. It felt like it added unnecessary complexity. Simplifying the trait felt like a win 🙏🏽.